src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,343 @@
+
+LogStream::LogStream() {
+    K2INFO("dtor");
+    _switched = false;    
+}
+
+LogStream::~LogStream() {
+    K2INFO("~dtor");
+}
+
+seastar::future<> 
+LogStream::init(String cpo_url, String persistenceClusrerName){
+    K2INFO("LogStream Init");
+    
+    _cpo = CPOClient(cpo_url);
+    return _client.init(persistenceClusrerName, cpo_url);
+}
+
+seastar::future<>
+LogStream::create(String logStreamName){
+    if (_logStreamCreate){
+        K2INFO("Log streaam already created");
+        throw k2::dto::LogStreamExistError("Log stream already created");
+    }
+
+    _logStreamName = std::move(logStreamName);
+    _logStreamCreate = true;
+
+    // create metadata plogs in advance
+    std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+    for (uint32_t i = 0; i < METADATA_PLOG_POOL_SIZE + 1; ++i){
+        createFutures.push_back(_client.create());
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        // register the metadata plog info to CPO
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for Metadata LogStream");
+            }
+            _metadataPlogPool.push_back(std::move(plogId));
+        }
+        String plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+        _metadataPlogPool.pop_back();
+        _metaInfo = std::move(std::make_pair(std::move(plogId), 0));
+        return _cpo.RegisterMetadataStreamLog(Deadline<>(_cpo_timeout()), _logStreamName, _metaInfo.first);
+    })
+    .then([this] (auto&& response){
+        // create WAL plogs in advance
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw k2::dto::LogStreamMetadataRegisterError("unable to register Metadata LogStream Plog to CPO");
+        }
+        std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+        for (uint32_t i = 0; i < WAL_PLOG_POOL_SIZE + 1; ++i){
+            createFutures.push_back(_client.create());
+        }
+        return seastar::when_all_succeed(createFutures.begin(), createFutures.end());
+    })
+    .then([this] (auto&& responses){
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for WAL");
+            }
+            _walPlogPool.push_back(std::move(plogId));
+        }
+        String plogId = _walPlogPool[_walPlogPool.size()-1];
+        _walPlogPool.pop_back();
+        _walInfo = std::move(std::make_pair(plogId, 0));
+
+        // write the metadata of the first WAL plog to metadata log streams
+        Payload temp_payload(Payload::DefaultAllocator);
+        temp_payload.write(std::move(plogId));
+        return _write(std::move(temp_payload), 0);

```

Comment:
second parameter better use "false" instead of 0.

Reply:
fixed